R/semla_to_seurat.R

```diff
@@ -29,63 +30,126 @@ NULL
 #' # Use Seurat functions
 #' SpatialFeaturePlot(se_merged, features = "Nrgn")
 #' 
+#' # Transform images in object
+#' se_merged <- LoadImages(se_merged)
+#' rotation_angle <- get_array_rotation(se, grid_pattern = "hexagonal")
+#' transforms <- rbind(generate_rigid_transform(sampleID = 1L, angle = rotation_angle$`1`),
+#'                     generate_rigid_transform(sampleID = 2L, angle = 12))
+#' se_merged <- RigidTransformImages(se_merged, transforms = transforms)
+#' 
+#' # Update object
+#' se_raw <- UpdateSeuratFromSemla(se_merged, image_use = "transformed")
+#' se_transformed <- UpdateSeuratFromSemla(se_merged, image_use = "raw")
+#' 
+#' # Use Seurat functions
+#' p1 <- SpatialFeaturePlot(se_raw, features = "Nrgn")
+#' p2 <- SpatialFeaturePlot(se_transformed, features = "Nrgn")
+#' p1 | p2
+#' 
 #' @export
 UpdateSeuratFromSemla <- function (
     object,
+    image_use = "raw",
     verbose = TRUE
 ) {
   
   if (verbose) cli_h3("Adding @images data from {col_br_magenta('Staffli')} object")
   if (verbose) cat_line()
   
   # Set global variables to NULL
-  pxl_row_in_fullres <- pxl_col_in_fullres <- sampleID <- NULL
+  pxl_row_in_fullres <- pxl_col_in_fullres <- sampleID <- pxl_row_in_fullres_transformed <- pxl_col_in_fullres_transformed <- NULL
   
   # Check object
   if (!inherits(object, what = "Seurat")) abort(glue("Invalid class '{class(object)}'. Expected a 'Seurat' object."))
   .check_seurat_object(object)
   
+  image_use <- match.arg(image_use, choices = c("raw", "transformed"))
+  
   # Get Staffli object
   st_object <- GetStaffli(object)
-  
+
   # Add step for test data
   if (any(st_object@imgs %in% c("mousebrain", "mousecolon"))) {
     object <- LoadImages(object, verbose = FALSE)
     st_object <- GetStaffli(object)
   }
   
+  # Check if the transformed image exists
+  if (!image_use %in% names(st_object@rasterlists))
+    abort(glue("Transformed images are not available in this object."))
+  
   # Check image paths
   for (impath in st_object@imgs) {
     if (!file.exists(impath)) {
       abort(glue("Invalid image path '{impath}'"))
     }
   }
   
-  # Load images
-  if (verbose) cli_alert_info("Loading images")
-  imgs <- lapply(st_object@imgs, function(im) {
-    ar <- image_read(im)[[1]] |> as.integer()
-    ar <- ar/max(ar)
-    return(ar)
-  })
+  # Load images if using raw. If transformed, reverse from raster to 3D numeric array
+  if (image_use == "raw") {
+    if (verbose) cli_alert_info("Loading images")
+    imgs <- lapply(st_object@imgs, function(im) {
+      ar <- image_read(im)[[1]] |> as.integer()
+      ar <- ar/max(ar)
+      return(ar)
+    })
+  } else if (image_use == "transformed") {
+    if (verbose) cli_alert_info("Converting hexadecimal colors to RGB")
+    imgs <- lapply(object@tools[["Staffli"]]@rasterlists[["transformed"]], function(im) {
```

Comment:
I think you can use `image_read` directly on the raster `im`?

Reply:
Does it accept the raster `im` as input? I thought that you could only give a filepath

Reply:
From `image_read`:

path: `a file, url, or raster object or bitmap array`  

:-)

Reply:
No need to change if it already works I guess!